src/trex/quality_control.py

```diff
@@ -0,0 +1,375 @@
+                               ax: plt.Axes = None,
+                               ignore_incomplete: bool = True) -> plt.Axes:
+    """Plot histogram of Hamming distance between CloneIDs. ignore_incomplete is
+     set to True by default and it removes incomplete CloneIDs."""
+    if ignore_incomplete:
+        molecules = molecules[~molecules.clone_id.str.contains('-|0')]
+    this_clone_ids = molecules.clone_id.unique()
+    hamming_distances = np.empty([len(this_clone_ids)] * 2)
+
+    def my_iter(clone_id_list):
+        for inds in combinations(np.arange(len(clone_id_list)), 2):
+            yield inds, clone_id_list[inds[0]], clone_id_list[inds[1]]
+
+    # Hamming distance function
+    def is_similar(args):
+        inds, s, t = args
+        if not ignore_incomplete:
+            bad_chars = {'-', '0'}
+            if bad_chars & set(s) or bad_chars & set(t):
+                # Remove suffix and/or prefix where sequences do not overlap
+                s = s.lstrip("-0")
+                t = t[-len(s):]
+                t = t.lstrip("-0")
+                s = s[-len(t):]
+                s = s.rstrip("-0")
+                t = t[: len(s)]
+                t = t.rstrip("-0")
+                s = s[: len(t)]
+
+        return inds, hamming_distance(s, t)
+
+    for ind, val in map(is_similar, my_iter(this_clone_ids)):
+        hamming_distances[ind[0], ind[1]] = val
+
+    vals = hamming_distances[np.triu_indices_from(hamming_distances, 1)]
+    ax = sns.histplot(vals, discrete=True, log=True, ax=ax)
+
+    ax.set_title('Hamming Distance Histogram')
+    ax.set_xlabel('Hamming Distance')
+
+    return ax
+
+
+def jaccard(cell_1: npt.ArrayLike, cell_2: npt.ArrayLike) -> float:
+    """Estimates the Jaccard similarity between two boolean vectors taken from
+    the UMI count matrix."""
+    return sum(np.logical_and(cell_1, cell_2)) / sum(np.logical_or(cell_1,
+                                                                   cell_2))
+
+
+def jaccard_similarity_matrix(umi_count: pd.DataFrame) -> npt.ArrayLike:
+    """Builds Jaccard similarity matrix from the UMI count matrix loaded as
+    pandas DataFrame.
+
+    Note: columns are cell IDs and first column is disregarded as it usually has
+    the index to CloneIDs"""
+    this_cell_ids = umi_count.columns[1:]
+    jaccard_matrix = np.zeros([len(this_cell_ids)] * 2)
+
+    # Iterator over combinations of cells
+    def my_iter(cell_id_list):
+        for inds in combinations(np.arange(len(cell_id_list)), 2):
+            clones_cell_1 = umi_count[cell_id_list[inds[0]]].values > 0
+            clones_cell_2 = umi_count[cell_id_list[inds[1]]].values > 0
+            yield inds, clones_cell_1, clones_cell_2
+
+    def my_jaccard(args):
+        inds, clones_cell_1, clones_cell_2 = args
+        return inds, jaccard(clones_cell_1, clones_cell_2)
+
+    for ind, val in map(my_jaccard, my_iter(this_cell_ids)):
+        jaccard_matrix[ind[0], ind[1]] = val
```

Comment:
This looks a bit complicated. If you iterate over the results of `map()` applied to an iterator, you can just as well do the iteration directly.

Would this work instead?
```suggestion
    for id1, id2 in combinations(this_cell_ids, 2):
        clones_cell_1 = umi_count[id1].values > 0
        clones_cell_2 = umi_count[id2].values > 0
        jaccard_matrix[id1, id2] = jaccard(clones_cell_1, clones_cell_2)
```

Note also that you can just give the objects directly to `combinations`, no need to generate indices.